drv/vsc7448/src/spi.rs

```diff
@@ -0,0 +1,276 @@
+impl Vsc7448Spi {
+    /// Reads from a VSC7448 register
+    pub fn read<T>(&self, reg: RegisterAddress<T>) -> Result<T, VscError>
+    where
+        T: From<u32>,
+    {
+        assert!(reg.addr >= 0x71000000);
+        assert!(reg.addr <= 0x72000000);
+        let addr = (reg.addr & 0x00FFFFFF) >> 2;
+        let data: [u8; 3] = [
+            ((addr >> 16) & 0xFF) as u8,
+            ((addr >> 8) & 0xFF) as u8,
+            (addr & 0xFF) as u8,
+        ];
+
+        // We read back 7 + padding bytes in total:
+        // - 3 bytes of address
+        // - Some number of padding bytes
+        // - 4 bytes of data
+        const SIZE: usize = 7 + SPI_PAD_BYTES as usize;
+        let mut out = [0; SIZE];
+        self.0.exchange(&data[..], &mut out[..])?;
+        let value = (out[SIZE - 1] as u32)
+            | ((out[SIZE - 2] as u32) << 8)
+            | ((out[SIZE - 3] as u32) << 16)
+            | ((out[SIZE - 4] as u32) << 24);
+
+        ringbuf_entry_masked!(
+            DEBUG_TRACE_SPI,
+            Trace::Read {
+                addr: reg.addr,
+                value
+            }
+        );
+        // The VSC7448 is configured to return 0x88888888 if a register is
+        // read too fast.  Reading takes place over SPI: we write a 3-byte
+        // address, then read 4 bytes of data; based on SPI speed, we may
+        // need to configure padding bytes in between the address and
+        // returning data.
+        //
+        // This is controlled by setting DEVCPU_ORG:IF_CFGSTAT.IF_CFG in
+        // Vsc7448::init(), then by padding bytes in the `out` arrays in
+        // [read] and [write].
+        //
+        // Therefore, we should only read "too fast" if someone has modified
+        // the SPI speed without updating the padding byte, which should
+        // never happen in well-behaved code.
+        //
+        // If we see this sentinel value, then we check
+        // DEVCPU_ORG:IF_CFGSTAT.IF_STAT.  If that bit is set, then the sentinel
+        // value _actually_ indicates an error (and not just an unfortunate
+        // coincidence).
+        if value == 0x88888888 {
+            // Panic immediately if we got an invalid read sentinel while
+            // reading IF_CFGSTAT itself.  This check also protects us from a
+            // stack overflow (by panicking instead).
+            let if_cfgstat = Vsc7448::DEVCPU_ORG().DEVCPU_ORG().IF_CFGSTAT();
+            if reg.addr == if_cfgstat.addr {
+                panic!(
+                    "Got invalid read sentinel value while reading IF_CFGSTAT"
+                );
+            }
+            // This read should _never_ fail for timing reasons because the
+            // DEVCPU_ORG register block can be accessed faster than all other
+            // registers (section 5.3.2 of the datasheet).
+            let v = match self.read(if_cfgstat) {
+                Err(e) => {
+                    panic!("Failed read while trying to check IF_STAT: {:?}", e)
```

Comment:
Nit: this is equivalent to

```rust
let v = self.read(if_cfgstat)
    .expect("Failed read while trying to check IF_STAT");
```

Reply:
I ended up reworking this whole section to return errors instead!

Reply:
... then switched the nested read failure to a panic, because it indicates something went _deeply_ wrong.